### DISCUSSION TITLE
Where is the best place for a beginner to get advice on ReactiveUI coding?

### DISCUSSION THREAD

**Discussion by GarryPatchett** [reactiveui/ReactiveUI#2797] [Q&A]
I’m just beginning to start using ReactiveUI in a non-trivial way but have come up against all kinds of problems, a lot of my own making since I don’t really know what I’m doing yet.
I’ve used it in a simple application but I need to learn more so I can use it better.
I’ve tried looking at the documentation on the web but I’m getting bogged down in lots of terminology which I don’t understand yet.
I’ve tried looking at the Camelotia sample application but the things I need to know are ‘blanketed’ by lots of technology which I don’t use so it’s difficult for me to pick out the things which I actually need.

Is there somewhere I can go to ask questions such as:

- What have I done wrong with this bit of code?
- How can I improve how this works in a Reactive way?
- Why isn’t this code working?
- When the documentation says <whatever>, what does it actually mean by that?
- What does this thing do and when should, and shouldn’t, I use it?

Sometimes I just need someone to quickly point me in the right direction so I don’t keep going round and round in circles following the wrong path entirely.

**Comment by glennawatson** [comment]
You can ask in the discussion, we do have a active community on slack where you can ask questions, it's newbie friendly so no bad questions

https://www.reactiveui.net/slack

Also if you're in a position I would consider buying the reactiveui book, it's still mostly relevant to modern rxui programming https://www.reactiveui.net/book

If your company is willing to pay we also do offer training courses.

**Comment by glennawatson** [comment]
What terminology are you finding intimidating btw?

I might go over some you might not be aware of.

ReactiveUI is built ontop of Reactive Extensions. It's a separate library.

Reactive Extensions provides extension methods that handle 'events'. I like think of it like Linq for Events where it provides linq operators to your events.

An IObservable is like your event that you can subscribe to. IObserver<T> is like your notification mechanism.

ReactiveUI is built ontop of Reactive Extensions and adds a number of concepts to make UI development a bit easier.

Some of those are handling knowing when a property has changed but in a way that is observable. So WhenAnyValue (which should be like WhenPropertyChanged), means give me an observable I can subscribe to when the property value changes.

We also have a lot of wrappers for making observables usable by the UI. One is a ObservableAsPropertyHelper, which helps you expose the latest value from an observable event as a property on your view model.

Another is DynamicData which allows you to create a `ReadOnlyObservableCollection<T>` from an observable and all the values that are emitted from that observable.

So when I say linq for events I really mean it like 

```cs

public string Name
{ 
   get => _name;
   set => this.RaiseAndSetIfChanged(ref _name, value); // Helper method to trigger INotifyPropertyChanged
var nameChangedObservable = this.WhenAnyValue(x => x.Name);

nameChangedObservable.Select(x => x.ToLower()).Where(x => x == "Garry").Subscribe(x => Message.Show('Hi I'm Garry");
```

so above, think of the Select as converting it to lower case, we only trigger the subscribe if the name is "Garry" (you notice the 'x' a lot of people just use that for the previous value in the Linq statement). And subscribe is what happens when the "Name" property is changed.

I hope this gets rid of some of the jargon and if there is any others you can think of please feel free to ask here or on slack.

**Reply by GarryPatchett** [reply]
Thanks for your quick reply.

I’ve never used Slack before but I’ll give it a go – I might have to hang out at the edges for a while before I get involved.
The book is a little expensive (understandably so) but I might try it if all else fails.
I’m learning this by myself so there’s no company to pay for things or anyone to turn to for a quick chat.

As for the terminology, I think the main problem I’m having is that the Handbook https://www.reactiveui.net/docs/handbook/ is ordered alphabetically, rather than in the order I need to know things, so I don’t know where to start. I.e. The first thing is Collections, which starts off talking about Dynamic Data for collection-based observables, source caches and Linq-to-objects, etc. It’s quite overwhelming for a beginner.

Because of that I often come up against things that are probably explained elsewhere but I don’t know where because I don’t know the correct order in which to read things or what each thing is a part of.

I managed to get a simple app published in the Microsoft Store which uses Reactive UI but I found myself getting tied up in knots a bit at times. And that’s why I decided that I want to try and learn how to do it properly rather than my string-and-sticky-tape approach.
Also, not knowing much about XAML and all the things that go into that probably isn’t helping me much, but that’s entirely my problem.

Anyway, thanks again for the answer. I’m sure this won’t be my last question.
